addons/point_of_sale/static/src/app/models/related_models.js

```diff
@@ -658,9 +678,26 @@ export function createRelatedModels(modelDefs, modelClasses = {}, opts = {}) {
                         const existingRecords = items.filter((record) =>
                             exists(comodelName, record.id)
                         );
-                        for (const record2 of [...linkedRecs]) {
-                            disconnect(field, record, record2);
+                        const inverse = inverseMap.get(field);
+
+                        for (const record2 of linkedRecs) {
+                            if (field.type === "many2many") {
+                                if (inverse && inverse.name in record2) {
+                                    removeItem(record2, inverse.name, record);
+                                }
+                            } else if (field.type === "one2many") {
+                                // Detach the back-reference many2one
+                                if (inverse && record2[inverse.name] === record) {
+                                    record2[inverse.name] = undefined;
+                                }
```

Comment:
We should use `removeItem` here.

Reply:
Thanks for the review! However, I can't use `removeItem` here since this is the whole point of the PR to begin with. It's to avoid calling removeItem inside a for loop due to the tail update for reindexing that happens in `removeItem`. This was causing the O(n^2) complexity which I'm trying to avoid here by breaking it into 2 O(n). 

This is why I'm rebuilding the index separately once later. Am I missing something that's causing a bug that's not caught by the tests? I am no expert on POS or JS for that matter so please let me know what you think!